### PR TITLE
--cookies FILE

### PR DESCRIPTION
**Pull request by anonieee**
Adds a kinda dirty way of loading a cookie file.
Not sure how well this works in wider circumstances, though

### CODE REVIEW

**Comment by theychx** [comment]
Cool! Can you provide a specific example of usage of this functionality? So I can test myself.

File: totalsize/total.py [hunk 1]
```diff
@@ -430,6 +431,10 @@ def cli():
             csv_path = Path(csv_file)
             write_to_csv(csv_path, gen_csv_rows([MOCK_ENTRY]))
             csv_path.unlink()
+        
+        if args.cookies:
+            global YTDL_OPTS
+            YTDL_OPTS['cookiefile'] = args.cookies
```
Comment (theychx) [inline totalsize/total.py]: `YTDL_OPTS` is a constant, so you're not supposed to modify it. Instead, you could add a `cookies` param to the initializer of `Playlist`, and then have a new (local) variable in the initializer, with the contents of `YTDL_OPTS` plus your cookies stuff. You'd have to modify the creation of the `playlist` object in `cli` accordingly of course.

Comment (anonieee) [reply totalsize/total.py]: Ok, I've tried fixing it.

**Comment by anonieee** [comment]
Well, my personal use case is that YouTube has been giving me 429 errors lately even though I haven't been downloading stuff all that much. Adding cookies fixed it for me.
This should also be useful if you want to access login-restricted videos.

**Comment by theychx** [comment]
What happens if the supplied cookies file is borked (or not a cookies file at all)?

File: totalsize/total.py [hunk 2]
```diff
@@ -139,7 +139,10 @@ def _readable_amount(self, amount, byte=False):
 
 
 class Playlist:
-    def __init__(self, url, format_sel, retries=0):
+    def __init__(self, url, format_sel, retries=0, cookies=None):
+        if cookies:
+            YTDL_OPTS['cookiefile'] = cookies
```
Comment (theychx) [inline totalsize/total.py]: Please make a new local variable, and then use that in the contruction of `self._ydl`.

Comment (anonieee) [reply totalsize/total.py]: When you're attempting to modify a global variable inside a function without using global VAR, wouldn't it automatically become a local variable?

**Comment by anonieee** [comment]
When the file doesn't exist it proceeds normally, but if the file is formatted incorrectly the program crashes with an exception.
I'm trying to find a way to verify the validity of a Netscape cookie file in Python, but so far I can't find any.

File: totalsize/total.py (hunk 2 above)
Comment (anonieee) [reply totalsize/total.py]: Or would something like this be better?

Comment (theychx) [reply totalsize/total.py]: Exactly. Please use a lowercase name (something like `opts` is fine). All-caps names are for constants.

**Comment by anonieee** [comment]
Added a validity check for the cookie file

File: totalsize/total.py (hunk 2 above)
Comment (theychx) [reply totalsize/total.py]:
```
opts = YTDL_OPTS
opts["cookiefile"] = cookies
```

File: totalsize/total.py [hunk 3]
```diff
@@ -139,9 +149,20 @@ def _readable_amount(self, amount, byte=False):
 
 
 class Playlist:
-    def __init__(self, url, format_sel, retries=0):
+    def __init__(self, url, format_sel, retries=0, cookies=None):
+        YTDL_OPTS_LOCAL = YTDL_OPTS
```
Comment (theychx) [inline totalsize/total.py]: Please use lowercase name i.e.:
```
opts = YTDL_OPTS
opts["cookiefile"] = cookies
```

File: totalsize/total.py [hunk 4]
```diff
@@ -139,9 +149,20 @@ def _readable_amount(self, amount, byte=False):
...
+        if cookies:
+            if not os.path.isfile(cookies):
+                raise CookieFileError  
+            try:
+                cj = http.cookiejar.MozillaCookieJar()
+                cj.load(cookies, ignore_discard=True, ignore_expires=True)
+            except:
```
Comment (theychx) [inline totalsize/total.py]: Catch-all except statements are not cool. Please investigate exactly what exception(s) are thrown when the cookie file is invalid, and then catch only those.

File: totalsize/total.py (hunk 3 above)
Comment (anonieee) [reply totalsize/total.py]: Fixed

File: totalsize/total.py (hunk 4 above)
Comment (anonieee) [reply totalsize/total.py]: Fixed, is this correct?

**Comment by theychx** [comment]
You forgot to push. :smiley:

**Comment by anonieee** [comment]
Whoops pushed to the wrong branch

**Comment by theychx** [comment]
Great work! Thanks!

**Pull request merged**